### ISSUE TITLE
Problem with -DefaultProfile in Az.Accounts 2.7.0

### ISSUE THREAD

**Issue by misterfoo** [Azure/azure-powershell#16617]
### Description

Something seems broken in the latest version of Az.Accounts; commands that were previously working are now failing with the error shown in the debug output below. The code had been working for a long time in pipelines which showed version 2.6.2 of Az.Accounts being used by the AzurePowerShell@5 task, and stopped working at the same time that version 2.7.0 of Az.Accounts started being used.

Running Get-AzResourceGroup with no `-DefaultProfile` argument works correctly. The service connection and service principal are working correctly and have valid credentials. Note that the usage of `-DefaultProfile` in the issue script below is a simplified version of the real code, which is library code that can't assume only a single subscription available in the Az context (i.e. we do need to use `-DefaultProfile` here).

### Issue script & Debug output

```PowerShell
<Note: this code is running in the context of an AzurePowerShell@5 task in a DevOps pipeline running on a windows-latest pool agent>

PS> $DebugPreference = "Continue"
PS> $ctx = Get-AzContext -ListAvailable | Select-Object -First 1
PS> Get-AzResourceGroup -Location eastus -DefaultProfile $ctx

"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\c92199f2-34c2-4fde-8e14-b874ba232f22.ps1'"
Added TLS 1.2 in session.
Import-Module -Name C:\Modules\az_6.5.0\Az.Accounts\2.7.0\Az.Accounts.psd1 -Global
WARNING: Both Az and AzureRM modules were detected on this machine. Az and AzureRM modules cannot be imported in the same session or used in the same script or runbook. If you are running PowerShell in an environment you control you can use the 'Uninstall-AzureRm' cmdlet to remove all AzureRm modules from your machine. If you are running in Azure Automation, take care that none of your runbooks import both Az and AzureRM modules. More information can be found here: https://aka.ms/azps-migration-guide
Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue
Clear-AzContext -Scope Process
Connect-AzAccount -ServicePrincipal -Tenant *** -Credential System.Management.Automation.PSCredential -Environment AzureCloud @processScope
 Set-AzContext -SubscriptionId 2274829b-0ad3-493a-a695-d5f3e1f99a2d -TenantId ***
DEBUG: 9:42:19 PM - GetAzureResourceGroupCmdlet begin processing with ParameterSet 'GetByResourceGroupName'.
DEBUG: 9:42:19 PM - using account id '***'...
DEBUG: [Common.Authentication]: Authenticating using Account: '***', environment: 'AzureCloud', tenant: '***'
DEBUG: [Common.Authentication]: Received exception No certificate thumbprint or secret provided for the given service principal '***'., while authenticating.
Get-AzResourceGroup: D:\a\_temp\5bce6fd3-fb0c-45e7-beeb-f9a8eccca2e5.ps1:38
Line |
  38 |  Get-AzResourceGroup -Location eastus -DefaultProfile $ctx
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Your Azure credentials have not been set up or have expired, please run Connect-AzAccount to set up
     | your Azure credentials. No certificate thumbprint or secret provided for the given service principal
     | '***'.

DEBUG: AzureQoSEvent: Module: Az.Resources:4.4.0; CommandName: Get-AzResourceGroup; PSVersion: 7.2.0; IsSuccess: False; Duration: 00:00:00.1532675; Exception: Your Azure credentials have not been set up or have expired, please run Connect-AzAccount to set up your Azure credentials.
No certificate thumbprint or secret provided for the given service principal '***'.;
DEBUG: Finish sending metric.
DEBUG: 9:42:19 PM - GetAzureResourceGroupCmdlet end processing.
```

### Environment data

```PowerShell
Key   : PSVersion
Value : 7.2.0
Name  : PSVersion

Key   : PSEdition
Value : Core
Name  : PSEdition

Key   : GitCommitId
Value : 7.2.0
Name  : GitCommitId

Key   : OS
Value : Microsoft Windows 10.0.17763
Name  : OS

Key   : Platform
Value : Win32NT
Name  : Platform

Key   : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0…}
Name  : PSCompatibleVersions

Key   : PSRemotingProtocolVersion
Value : 2.3
Name  : PSRemotingProtocolVersion

Key   : SerializationVersion
Value : 1.1.0.1
Name  : SerializationVersion

Key   : WSManStackVersion
Value : 3.0
Name  : WSManStackVersion
```

### Module versions

```PowerShell
Name              : Az.Accounts
Path              : C:\Modules\az_6.5.0\Az.Accounts\2.7.0\Az.Accounts.psm1
Description       : Microsoft Azure PowerShell - Accounts credential management cmdlets for Azure Resource Manager in 
                    Windows PowerShell and PowerShell Core.
                    
                    For more information on account credential management, please visit the following: 
                    https://docs.microsoft.com/powershell/azure/authenticate-azureps
Guid              : 17a2feff-488b-47f9-8729-e2cec094624c
Version           : 2.7.0
ModuleBase        : C:\Modules\az_6.5.0\Az.Accounts\2.7.0
ModuleType        : Script
PrivateData       : {PSData}
AccessMode        : ReadWrite
ExportedAliases   : {[Add-AzAccount, Add-AzAccount], [Get-AzDomain, Get-AzDomain], [Invoke-AzRest, Invoke-AzRest], 
                    [Login-AzAccount, Login-AzAccount]…}
ExportedCmdlets   : {[Add-AzEnvironment, Add-AzEnvironment], [Clear-AzContext, Clear-AzContext], [Clear-AzDefault, 
                    Clear-AzDefault], [Connect-AzAccount, Connect-AzAccount]…}
ExportedFunctions : {}
ExportedVariables : {}
NestedModules     : {Microsoft.Azure.PowerShell.Cmdlets.Accounts}


Name              : Az.Resources
Path              : C:\Modules\az_6.5.0\Az.Resources\4.4.0\Az.Resources.psm1
Description       : Microsoft Azure PowerShell - Azure Resource Manager and Active Directory cmdlets in Windows 
                    PowerShell and PowerShell Core.  Manages subscriptions, tenants, resource groups, deployment 
                    templates, providers, and resource permissions in Azure Resource Manager.  Provides cmdlets for 
                    managing resources generically across resource providers.
                    
                    For more information on Resource Manager, please visit the following: 
                    https://docs.microsoft.com/azure/azure-resource-manager/
                    For more information on Active Directory, please visit the following: 
                    https://docs.microsoft.com/azure/active-directory/fundamentals/active-directory-whatis
Guid              : 48bb344d-4c24-441e-8ea0-589947784700
Version           : 4.4.0
ModuleBase        : C:\Modules\az_6.5.0\Az.Resources\4.4.0
ModuleType        : Script
PrivateData       : {PSData}
AccessMode        : ReadWrite
ExportedAliases   : {[Get-AzADServicePrincipalCredential, Get-AzADServicePrincipalCredential], 
                    [Get-AzResourceProviderAction, Get-AzResourceProviderAction], [Get-AzSubscriptionDeployment, 
                    Get-AzSubscriptionDeployment], [Get-AzSubscriptionDeploymentOperation, 
                    Get-AzSubscriptionDeploymentOperation]…}
ExportedCmdlets   : {[Add-AzADGroupMember, Add-AzADGroupMember], [Export-AzResourceGroup, Export-AzResourceGroup], 
                    [Export-AzTemplateSpec, Export-AzTemplateSpec], [Get-AzADAppCredential, Get-AzADAppCredential]…}
ExportedFunctions : {}
ExportedVariables : {}
NestedModules     : {Microsoft.Azure.PowerShell.Cmdlets.Resources, Microsoft.Azure.PowerShell.Cmdlets.ResourceManager, 
                    Microsoft.Azure.PowerShell.Cmdlets.Tags}
```

### Error output

```PowerShell
Message        : Your Azure credentials have not been set up or have expired, please run Connect-AzAccount to set up 
                 your Azure credentials.
                 No certificate thumbprint or secret provided for the given service principal 
                 '***'.
StackTrace     :    at Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.ResourceManagerCmdletBase.Handle
                 Exception(ExceptionDispatchInfo capturedException)
                    at Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.ResourceManagerCmdletBase.Execut
                 eCmdlet()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.<>c__3`1.<ExecuteSynchronously
                 OrAsJob>b__3_0(T c)
                    at 
                 Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.ExecuteSynchronouslyOrAsJob[T](T 
                 cmdlet, Action`1 executor)
                    at 
                 Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.ExecuteSynchronouslyOrAsJob[T](T 
                 cmdlet)
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
Exception      : Microsoft.Azure.Commands.Common.Exceptions.AzPSArgumentException
InvocationInfo : {Get-AzResourceGroup}
Line           : Get-AzResourceGroup -Location eastus -DefaultProfile $ctx
                 
Position       : At D:\a\_temp\c92199f2-34c2-4fde-8e14-b874ba232f22.ps1:38 char:1
                 + Get-AzResourceGroup -Location eastus -DefaultProfile $ctx
                 + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
HistoryId      : 1


Message        : No certificate thumbprint or secret provided for the given service principal 
                 '***'.
StackTrace     :    at Microsoft.Azure.PowerShell.Authenticators.ServicePrincipalAuthenticator.Authenticate(Authenticat
                 ionParameters parameters, CancellationToken cancellationToken)
                    at Microsoft.Azure.Commands.Common.Authentication.DelegatingAuthenticator.TryAuthenticate(Authentic
                 ationParameters parameters, CancellationToken cancellationToken, Task`1& token)
                    at Microsoft.Azure.Commands.Common.Authentication.DelegatingAuthenticator.TryAuthenticate(Authentic
                 ationParameters parameters, CancellationToken cancellationToken, Task`1& token)
                    at Microsoft.Azure.Commands.Common.Authentication.DelegatingAuthenticator.TryAuthenticate(Authentic
                 ationParameters parameters, CancellationToken cancellationToken, Task`1& token)
                    at Microsoft.Azure.Commands.Common.Authentication.DelegatingAuthenticator.TryAuthenticate(Authentic
                 ationParameters parameters, CancellationToken cancellationToken, Task`1& token)
                    at Microsoft.Azure.Commands.Common.Authentication.DelegatingAuthenticator.TryAuthenticate(Authentic
                 ationParameters parameters, Task`1& token)
                    at Microsoft.Azure.Commands.Common.Authentication.Factories.AuthenticationFactory.Authenticate(IAzu
                 reAccount account, IAzureEnvironment environment, String tenant, SecureString password, String 
                 promptBehavior, Action`1 promptAction, IAzureTokenCache tokenCache, String resourceId)
                    at Microsoft.Azure.Commands.Common.Authentication.Factories.AuthenticationFactory.Authenticate(IAzu
                 reAccount account, IAzureEnvironment environment, String tenant, SecureString password, String 
                 promptBehavior, Action`1 promptAction, String resourceId)
                    at Microsoft.Azure.Commands.Common.Authentication.Factories.AuthenticationFactory.GetServiceClientC
                 redentials(IAzureContext context, String targetEndpoint, String resourceId)
Exception      : Microsoft.Identity.Client.MsalException
InvocationInfo : {Get-AzResourceGroup}
Line           : Get-AzResourceGroup -Location eastus -DefaultProfile $ctx
                 
Position       : At D:\a\_temp\c92199f2-34c2-4fde-8e14-b874ba232f22.ps1:38 char:1
                 + Get-AzResourceGroup -Location eastus -DefaultProfile $ctx
                 + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
HistoryId      : 1
```

**Comment by dingmeng-xue** [comment]
Thanks for reporting. @msJinLei , please look into this issue. Is it due to secret is not merged into context when context is passed in as parameter

**Comment by waltervos** [comment]
I'm experiencing the same issue, it seems. My code is: 
```
$Context = Get-AzContext
$Token = [Microsoft.Azure.Commands.Common.Authentication.AzureSession]::Instance.AuthenticationFactory.Authenticate($Context.Account, $Context.Environment, $Context.Tenant.Id.ToString(), $null, [Microsoft.Azure.Commands.Common.Authentication.ShowDialog]::Never, $null, "https://westeurope.asazure.windows.net")
```

This is some often copy-pasted code that is used to retrieve an access token in Azure Pipelines. Is there a workaround for this?

**Comment by dingmeng-xue** [comment]
Hi @waltervos , 

Azure PowerShell team minimizes breaking change based on cmdlet signature. If the user call internal lib and methods, we cannot ensure it always works. 

For your case, I believe you need to create a new Azure environment and login with your account. https://docs.microsoft.com/en-us/powershell/module/az.accounts/add-azenvironment?view=azps-7.0.0

**Comment by waltervos** [comment]
Thanks @dingmeng-xue, sounds like my team should be adhering to "don't call internal lib and methods" :). My use case to obtain the access token, which I found out is trivial to do by just calling `Get-AzAccessToken` and using the Token attribute of the returned value. Thanks!

**Comment by mschoneman** [comment]
We were able to work around this by changing `inputs` in our `AzurePowerShell` task in Azure DevOps :

from:
```
     azurePowerShellVersion: latestVersion
```
to
```
      azurePowerShellVersion: otherVersion
      preferredAzurePowerShellVersion: 6.6.0
```

**Comment by ghvanderweg** [comment]
I believe this issue is caused by https://github.com/Azure/azure-powershell/pull/16449.
Any chance that could be reverted, since that PR inadvertently caused a breaking change?

**Comment by dingmeng-xue** [comment]
Hi @ghvanderweg , that breaking change is by design in Az 7.0 release. `-DefautProfile` should be covered during that change. We are fixing it.

**Comment by ReneRebsdorf** [comment]
FYI, it seems this only occurs when working with service principals.
For Azure Contexts originating from a 'regular user', I can not reproduce the problem.
With a service principal it is always reproducible.

**Comment by ReneRebsdorf** [comment]
This is reproducible outside of Azure DevOps by performing:

``` pwsh
Connect-AzAccount -ServicePrincipal -Credential $cred -Tenant "a-b-c-d"
$ctx = Select-AzContext "name-of-an-azure-context"
Get-AzResourceGroup -DefaultProfile $ctx
```

**Comment by dingmeng-xue** [comment]
> FYI, it seems this only occurs when working with service principals. For Azure Contexts originating from a 'regular user', I can not reproduce the problem. With a service principal it is always reproducible.

Correct. We hide secret of SP from profile. It's the reason why it is blocked when user use profile as parameter because it doesn't contain secret. We should load secret according to profile passed in.

**Comment by ReneRebsdorf** [comment]
> > FYI, it seems this only occurs when working with service principals. For Azure Contexts originating from a 'regular user', I can not reproduce the problem. With a service principal it is always reproducible.
> 
> Correct. We hide secret of SP from profile. It's the reason why it is blocked when user use profile as parameter because it doesn't contain secret. We should load secret according to profile passed in.

Do you have an estimate on when this will be fixed?
This breaks several of our client delivered pwsh modules, and I need to plan for implementering a hotfix by requiring the older version on January 3rd, and git revert the conversion changes that followed this update. That is unless a fix is expected to be delivered this year?

**Comment by msJinLei** [comment]
> > > FYI, it seems this only occurs when working with service principals. For Azure Contexts originating from a 'regular user', I can not reproduce the problem. With a service principal it is always reproducible.
> > 
> > 
> > Correct. We hide secret of SP from profile. It's the reason why it is blocked when user use profile as parameter because it doesn't contain secret. We should load secret according to profile passed in.
> 
> Do you have an estimate on when this will be fixed? This breaks several of our client delivered pwsh modules, and I need to plan for implementering a hotfix by requiring the older version on January 3rd, and git revert the conversion changes that followed this update. That is unless a fix is expected to be delivered this year?

Our next release in 01/04/2022 https://github.com/Azure/azure-powershell/milestone/130. We will include the fix into it.

There is a workaround @ReneRebsdorf @misterfoo. 
```
$current = Get-AzContext
$contexts = Get-AzContext -ListAvailable
Select-AzContext -Name $contexts[1].Name
.....do you jobs....
Select-AzContext -Context $current
```
It works if you don't run several Azure PowerShell jobs in parallel

**Comment by msJinLei** [comment]
We remove service principal secret from PSAzureRmAccount because customers can always find a way to get its value if we keep it in the object.
As a result, the context you retrieve from Get-AzContext doesn't have the secret. When the context is to set to DefaultProfie of  a new cmdlet. The cmdlet doesn't have enough information to get access token.
We handle the case of Set-AzContext and Select-AzContext but -DefaultProfile case is not covered in 2.7.0.

**Comment by ReneRebsdorf** [comment]
> We remove service principal secret from PSAzureRmAccount because customers can always find a way to get its value if we keep it in the object. As a result, the context you retrieve from Get-AzContext doesn't have the secret. When the context is to set to DefaultProfie of a new cmdlet. The cmdlet doesn't have enough information to get access token. We handle the case of Set-AzContext and Select-AzContext but -DefaultProfile case is not covered in 2.7.0.

@msJinLei 

Thank you for the update. Unfortunately the workaround is not useful (for us at least), as we use the profiles to accomplish parallel execution against multiple subscriptions.
My guess is this is the case for many others who use DefaultProfiles as it is the recommended method as per other GitHub Issues. Can't speak for @misterfoo and others, for whom the workaround may be useful.

I'll test this thoroughly on Jan. 4th.

**Comment by msJinLei** [comment]
Hi all (@ReneRebsdorf @misterfoo)
Please try this way to solve your issue, that is, to add back you service principal secret in your script

```
PS C:\Users\leijin> $context = Get-AzContext
PS C:\> $context.Account.ExtendedProperties.Add("ServicePrincipalSecret", "$yourServicePrincipalSecret")
```

**Comment by ReneRebsdorf** [comment]
@msJinLei How do you suggest to use this fix in Azure DevOps?

We currently create several threads, each which use their own Azure Context to target a subscription, which we create by iterating over subscriptions found with Get-AzSubscription.

The pipelines do not know the secret when executing, [unless using some 'workaround', which I belive is not an intended feature](https://pascalnaber.wordpress.com/2020/01/04/backdoor-in-azure-devops-to-get-the-password-of-a-service-principal/)

So the problem with the above comment is that it requires the pipelines to know the secret in order to add the key:value to ExtendedProperties.
